In [1]:
# If you what use ngrams.py you should use python2
# Or, otherwise, you need to modify ngrams.py by yourself in order to use it in python3.
from __future__ import print_function
from __future__ import unicode_literals

import re
import numpy as np
from ngrams import ngrams
from collections import defaultdict
from math import log, pow, inf

In [2]:
small_corpus = ['Why dont we start here',
                  'Why dont we end there',
                  'Let us start with a few other examples',
                  'We never start with an example with so few tokens',
                  'Tokens can be words that we start with in example docs'
                  ]

# HW3: Language Modeling
For this part of the assignment, you will implement two simple count-based n-gram language models: one based on maximum-likelihood estimation, and another based on Witten-Bell smoothing. The data you will be using is a subset of the Penn Treebank's tagged Wall Street Journal articles on which we have done some initial processing. There are two versions of the data for this assignment:

##### wsj.pos.gz
##### wsj-normalized.pos.gz
The difference is that, in the second (normalized) version of the data, we have collapsed some entries from certain tag categories (e.g. CDs, NNPs, etc.) into type-tokens to help reduce sparsity. Take a look at the data and see for yourself. Consider: what would be the benefits and drawbacks to this method of sparsity reduction? Note that, for this part of the assignment, the tags are un-necessary, so you'll want to work with the un-normalized version of the corpus.

https://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html

### Task 1: produce a tag-free corpus

For this task, you have two jobs. 
* First, you need to write a function to filter out all tags. 
* Second, Make sure your code works for both wsj.pos.gz and wsj-normalized.pos.gz

####What to turn in
* your code
* some samples to show me that your code works as it should be

In [3]:
import gzip

#regex = re.compile(r"/[A-Z$,.]{1,4}") #stackoverflow says unnecessary, cached anyway
maxln = 7

def rem_tag(line):
    line = line.decode('utf-8')
    return re.sub(r"/[A-Z$,.`']{1,4}",'',line)

print("######################## WSJ.POS Examples########################")
with gzip.open('wsj.pos.gz','r') as wsj:
    count = 0
    for line in wsj:
        print(line)
        print("\n")
        line = rem_tag(line)
        print(line)
        count += 1
        if count > maxln:
            break

print("######################## WSJ.NORMALIZED Examples:########################")
with gzip.open('wsj-normalized.pos.gz','r') as wsj:
    count = 0
    for line in wsj:
        print(line)
        print("\n")
        line = rem_tag(line)
        print(line)
        count += 1
        if count > maxln:
            break


######################## WSJ.POS Examples########################
b'Digital/NNP Equipment/NNP Corp./NNP reported/VBD a/DT 32/CD %/NN decline/NN in/IN net/JJ income/NN on/IN a/DT modest/JJ revenue/NN gain/NN in/IN its/PRP$ fiscal/JJ first/JJ quarter/NN ,/, causing/VBG some/DT analysts/NNS to/TO predict/VB weaker/JJR results/NNS ahead/RB than/IN they/PRP had/VBD expected/VBN ./.\n'


Digital Equipment Corp. reported a 32 % decline in net income on a modest revenue gain in its fiscal first quarter , causing some analysts to predict weaker results ahead than they had expected .

b'Although/IN the/DT second-largest/JJ computer/NN maker/NN had/VBD prepared/VBN Wall/NNP Street/NNP for/IN a/DT poor/JJ quarter/NN ,/, analysts/NNS said/VBD they/PRP were/VBD troubled/VBN by/IN signs/NNS of/IN flat/JJ U.S./NNP orders/NNS and/CC a/DT slowdown/NN in/IN the/DT rate/NN of/IN gain/NN in/IN foreign/JJ orders/NNS ./. The/DT Maynard/NNP ,/, Mass./NNP ,/, company/NN is/VBZ in/IN a/DT transition/NN in/IN wh

<font color="red">Self assessment:</font>

In [286]:
#Constructing Corpus
wsj_corpus = []
wsj_norm_corpus = []
with gzip.open('wsj.pos.gz','r') as wsj:
    for line in wsj:
        line = rem_tag(line)
        lines = line.replace(',','').split(' .')
        for sent in lines:    
            if len(sent) > 3:
                wsj_corpus += [sent.lower()]

with gzip.open('wsj-normalized.pos.gz','r') as wsj:
    for line in wsj:
        line = rem_tag(line)
        lines = line.replace(',','').split(' .')
        for sent in lines:    
            if len(sent) > 3:
                wsj_norm_corpus += [sent.lower()]


In [287]:
for i in range(6):
    print("WSJ:", wsj_corpus[i])
    print("WSJ NORM: ",wsj_norm_corpus[i])
print(len(wsj_corpus))
print(len(wsj_norm_corpus))


WSJ: digital equipment corp. reported a 32 % decline in net income on a modest revenue gain in its fiscal first quarter  causing some analysts to predict weaker results ahead than they had expected
WSJ NORM:  <nnp> <nnp> <nnp> reported a <cd> % decline in net income on a modest revenue gain in its fiscal first quarter  causing some analysts to predict weaker results ahead than they had expected
WSJ: although the second-largest computer maker had prepared wall street for a poor quarter  analysts said they were troubled by signs of flat u.s. orders and a slowdown in the rate of gain in foreign orders
WSJ NORM:  although the second-largest computer maker had prepared <nnp> <nnp> for a poor quarter  analysts said they were troubled by signs of flat <nnp> orders and a slowdown in the rate of gain in foreign orders
WSJ:  the maynard  mass.  company is in a transition in which it is trying to reduce its reliance on mid-range machines and establish a presence in workstations and mainframes
WSJ

### Maximum Likelihood
Now, start by producing code to compute maximum-likelihood estimate probabilities. Your code should be configurable with respect to the n-gram order- i.e., you should be able to set it to compute bigram, trigram, 4-gram, etc. probabilities. Refer to J&M and the lecture slides for definitions as needed. If you would like to write your own n-gram tokenization code, feel free to do so, but you may also use the ngrams.py utility class which contains a routine to take a list of tokens and produce a stream of n-grams with appropriate padding for the start and end of sentences.

#### Tip: 
* Start with a very small "toy" corpus of just a couple of sentences for debugging. 

* As discussed in class, I strongly recommend using nested defaultdicts as the foundational data structure for your language model, where the "outer" key is the prefix, and the value retrieved by that prefix is a second defaultdict  containing possible suffices for that prefix, each of which is an "inner" key. E.g., p("TRUTHS" | "HOLD THESE") would be retrieved by first looking up "HOLD THESE" and then from the resulting dictionary, looking up "TRUTHS": prob = trigrams[("HOLD","THESE")]["TRUTHS"] . Note that this arrangement makes it very easy to e.g. find out the number of times a given history occurs, the total probability mass assigned to all of a history's continuations, etc., all of which will be extremely helpful in the next part of the assignment.

* Use tuples to represent prefixes. E.g., instead of the string "HOLD THESE", use the tuple ("HOLD", "THESE"). Note that, in Python, lists are mutable, and therefore may not be used as keys in dictionaries- but tuples are immutable, and so make excellent keys.

* Don't forget about numerical underflow issues! You'll want to represent probabilities as negative base-2 log probabilities, and modify your arithmetic accordingly. I recommend experimenting with [the bitweight Python library](https://github.com/stevenbedrick/bitweight) (see its unit tests for example usage).
* 

#### What to turn in:
* your code 
* use your code to create a simple language model for small_corpus named as small_lm and show me that your output is correct(This is a small coupus so you could manully calculate the probalility).
* use your code to create language model for wsj.pos.gz named as wsj_lm

In [168]:
def log2(x):
    """Return -log2(x)"""
    if x == 0:
        return inf
    return -log(x, 2)

def pow2(x):
    """Return 2^-x"""
    return pow(2,-x)

def make_dict(corpus, order):
    """Returns a nested defaultdict(defaultdict), formatted as follows:
        [prefix]:[suffix]:prefix/suffix count"""
    corpus_list = []
    corpus_dict = defaultdict(lambda: defaultdict(int))

    corpus_list += [corpus[i].lower().split() for i in range(len(corpus))] #corpus[i].lower().split() ?
    chunks = ngrams(corpus_list, order)
    lchunks = list(chunks)
    
    for i in range(len(lchunks)):
        corpus_dict[lchunks[i][0] ][lchunks[i][1]] += 1

    return corpus_dict

def mle_solo(corp_dict, entry):
    """single n-gram probability function
    Input: dictionary(dictionary()): corp_dict, String: sentence
    Returns: ngrams as a list
    Order identified from dictionary used"""

    if entry[1] not in corp_dict[entry[0]]:
        return 0
    total = 0
    for val in corp_dict[entry[0]]:
        total += corp_dict[entry[0]][val]

    #prob = corp_dict[entry[0]][entry[1]] / total
    prob = log2(corp_dict[entry[0]][entry[1]]) - log2(total)
    prob = pow2(prob)
    return prob

def mle100(corp_dict, sentence):
    """Multiply version of MLE, whole sentence (i.e. parsed through ngrams() for <S_0>, etc)"""
    acc = 1
    order = len(list(corp_dict.keys())[0]) + 1
    n = list(ngrams([sentence.split()], order))
    #print(n)
    for entry in n:
        if entry[1] not in corp_dict[entry[0]]:
            return 0
        total = 0
        for i in corp_dict[entry[0]]:
            total += corp_dict[entry[0]][i]
        #print(total)
        prob = corp_dict[entry[0]][entry[1]] / total
        acc = acc * prob
        #print("acc@", entry[0], ":", entry[1] ,acc)
    return acc

def mle_log(corp_dict, sentence):
    """Log version of MLE, returns in -log2"""
    sentence = sentence.lower()
    acc = 0 #in log2
    order = len(list(corp_dict.keys())[0]) + 1
    n = list(ngrams([sentence.split()], order))
    for entry in n:
        total = 0
        for i in corp_dict[entry[0]]:
            total += corp_dict[entry[0]][i]
        #print("total",total)
        if entry[1] not in corp_dict[entry[0]]:
            return inf
        suffix = corp_dict[entry[0]][entry[1]]
        if (suffix == 0):
            return inf
        prob = log2(suffix) - log2(total)
        #print("PROB", prob)
        assert prob >= 0, "-log2(prob) not > 0" #since negative log(fraction)
        acc = acc + prob
        #print("acc@", entry[0], ":", entry[1] ,acc)
    return acc

def print_dict(corpus_dict):
    keys = corpus_dict.keys()
    
    for key in sorted(keys):
        print(key)
        for key2 in sorted(corpus_dict[key].keys()):
            print("\t", key2, ":", corpus_dict[key][key2])

### Small Toy Test Case, BIGRAMS
small_lm = make_dict(small_corpus, 2)

### MLE with multiplication
print("Bigram (0 case): <start>:", mle100(small_lm, "start")) #0-probability case
print("mle100", mle100(small_lm, "we start here")) #non-zero probability case

### MLE with log
res1 = mle_log(small_lm, "start")
print("=0 until smoothing: log: ", res1, "decimal:", pow2(res1)) #0-probability case
res2 = mle_log(small_lm, "we start here") #non-zero probability case
print("mle log: ", res2, "decimal prob:", pow2(res2))

res3 = mle_log(small_lm, "Why dont we end there")
print("'why dont we end there':mle log: ", res3, "decimal prob:", pow2(res3))

small_lm1 = make_dict(small_corpus, 1)
testword = "us"
res = mle_log(small_lm1, testword)
print("unigram: ",testword, res, "decimal:", pow2(res))

### Trigram
testword = "why dont we start here"
res = mle_log(make_dict(small_corpus, 3), testword)
print("trigram: ",testword, res, "decimal:", pow2(res))


Bigram (0 case): <start>: 0
mle100 0.025
=0 until smoothing: log:  inf decimal: 0.0
mle log:  5.321928094887362 decimal prob: 0.025000000000000012
'why dont we end there':mle log:  3.321928094887362 decimal prob: 0.1
unigram:  us 5.285402218862249 decimal: 0.025641025641025633
trigram:  why dont we start here 3.321928094887362 decimal: 0.1


In [270]:
wsj_lm = make_dict(wsj_corpus, 2)
wsj_norm_lm = make_dict(wsj_norm_corpus, 2)
print(len(wsj_lm.keys()), len(wsj_norm_lm.keys()))

43754 29006


In [9]:
res = mle_log(wsj_lm, "In New York") #non-zero probability case
print("mle (negative) log: ", res, "decimal prob:", pow2(res))

res = mle_log(wsj_lm, "Pineapple") #zero probability case
print("mle (negative) log: ", res, "decimal prob:", pow2(res))

mle (negative) log:  16.0467925383373 decimal prob: 1.4771823564213143e-05
mle (negative) log:  inf decimal prob: 0.0


<font color="red">Self assessment:</font>

### Smoothing

Once you’ve got an unsmoothed model working, move on to implementing Witten-Bell smoothing. Refer to the slides and J&M for details on how that ought to work.

#### Tip: 
* You can modify an already-populated defaultdict to change its default value (for example, to store a default backoff value for a particular history) by changing the object’s default_factory attribute. Consult the documentation for examples of how this works.
* As defined, W-B smoothing is highly recursive; you may find it more efficient to re-frame the algorithm in iterative terms.
* As in the previous section, start small.
* [This may offer you some help on how to implement Witten-Bell smoothing](http://www.ee.columbia.edu/~stanchen/e6884/labs/lab3/x207.html)


#### What to turn in:
* your code 
* use your code to create a simple smoothed language model based on small_lm  and show me that your output is correct(This is a small coupus so you could manully calculate the probalility).
* use your code to create a smoothed language model based on wsj_lm

In [10]:
#Default value usage test>Pass
d1 = defaultdict(int)
d1["apple"] = 5
print(d1[5])
x = 6
d1.default_factory = lambda x=x: x #6
print(d1.default_factory(), d1[100], d1[5], d1["apple"]) #Result 6 6 0 5
#Note: d1[5] set to initial value 0 after first call
#Note: Use d1.default_factory() to call lambda function directly without creating new entry

d2 = defaultdict(lambda: defaultdict(int))
d2[1][2] = 15
x = 0.5
d2[1].default_factory = lambda x=x: x #0.5
print(d2[1][2], d2[1][111], d2[2][2]) #Expected 15, 0.5, 0

#Note: Changing lambda: 6 to lambda: x causes an issue.
#https://stackoverflow.com/questions/10452770/python-lambdas-binding-to-local-values
#https://stackoverflow.com/questions/21053988/lambda-function-acessing-outside-variable
print(d1.default_factory(), d1[100], d1[5], d1["apple"])
print(d2[1][2], d2[1][111], d2[2][2])

#Note: Use lambda x=x: x to capture

0
6 6 0 5
15 0.5 0
6 6 0 5
15 0.5 0


In [300]:
def make_lamb_dicts(models, n, corpus, debug = 0):
    #Initialize n-gram dictionaries from unigram to desired n-gram
    for i in range(n):
        lm = make_dict(corpus, i+1)
        models.append(lm.copy())
    #Verify
    #print_dict(models[n-1])

    #Build up lambda terms for unigram to (n)-gram, storing in models[]
    for i in range(0,n):
        keys = models[i].keys()
        if debug:
            print("KEYS: ", keys)
        for key in keys:
            inner_keys = models[i][key].keys()
            if debug:
                print("INNER KEYS: ", inner_keys)

            tokens = 0
            for ikey in inner_keys:
                tokens += models[i][key][ikey]

            types = 0
            types = len(inner_keys)
            
            if debug:
                print("(tokens types)", tokens, types)
            lamb = tokens / (tokens+types)
            models[i][key].default_factory = lambda lamb=lamb: lamb


In [145]:
n = 3 #up to trigrams created
testlambs = []
make_lamb_dicts(testlambs, n, small_corpus, debug = 0)
#print_dict(testlambs[1])

In [237]:
#Calculate Smoothed Probability, starting with unigram level (iterative)
#list_grams = list(ngrams([test_case.split()], 2))
#print(list_grams)

def wbSmooth(list_dicts, sentence, order, debug = 0):
    """
    Input:
        list_dicts: List of nested defaultdict(defaultdict(int)) structure containing n-gram counts
            from 1-gram to (order)-gram.
        sentence: Input sentence to calculate probability on
        order: degree of n-gram. Must be within provided (list_dict) range
    Output: Decimal WB smoothed probability of input
    """
    sentence=sentence.lower()

    uni_tokens = sum(list_dicts[0][()].values())
    uni_types = len(list_dicts[0][()].keys())
    lamb_dot = uni_tokens / (uni_tokens+uni_types)
    unseen = 1/(uni_tokens+uni_types)
    if debug:
        print("uni_tokens, uni_types:", uni_tokens, uni_types)
        print("lambda.dot = ", lamb_dot)
        print("unseen", unseen)
    
    grams = list(ngrams([sentence.split()], order))
    if debug:
        print(order,"-grams:", grams)
    probs = [] #WB probabilities of component n-grams of the input

#Calculate smoothed prob for every n-gram in (sentence) and add to accumulator
    for gram in grams:
        prefix = gram[0]
        suffix = gram[1]
        if debug:
            print("current gram: ", gram)

        ptemp =0.5

        for i in range(0,order):
            if i == 0: #base case, unigrams
                #p = lamb_dot * pow2(mle_log(list_dicts[i], suffix)) + unseen
                p = lamb_dot * (mle100(list_dicts[i], suffix)) + unseen
                #p = pow2( log2(lamb_dot) + log2((mle100(list_dicts[i], suffix))) ) + unseen #no effect
                ptemp = p
            else:
                gram_dict = list_dicts[i][prefix[order-i-1:]] #slice of prefix for current level of n-gram
                lamb = gram_dict.default_factory()
                s = [prefix[order-i-1:], suffix]
                mle = mle_solo(list_dicts[i], s)
                ptemp = lamb*mle + (1-lamb)*ptemp
                #ptemp = pow2(log2(lamb)+log2(mle)) + pow2(log2((1-lamb))+log2(ptemp)) #no effect
                
                if debug:
                    print("s: ",s, "ptemp: ", ptemp)
                    print("mle", s, mle_solo(list_dicts[i], s))
            
            #print("prefix", (gram[order-2-i],) ) #Unigram prefix
        probs.append( ptemp)
#Multiply (log form) individual gram probabilities together
    ret = 0
    for prob in probs:
        ret += log2(prob)
    return pow2(ret)

test_case = "let"
print("unigram smoothed: ",test_case, wbSmooth(testlambs ,test_case, 1))
test_case = "let us"
print("bigram smoothed: ",test_case, wbSmooth(testlambs ,test_case, 2, debug=1))
test_case = "why why"
print("bigram unseen pair smoothed: ",test_case, wbSmooth(testlambs ,test_case, 2, debug=0))
test_case = "us with"
print("bigram unseen pair smoothed: ",test_case, wbSmooth(testlambs ,test_case, 2, debug=0))
test_case = "chicken chicken"
print("bigram unseen tokens smoothed: ",test_case, wbSmooth(testlambs ,test_case, 2, debug=0))

testword = "let"
res = mle_log(small_lm1, testword)
print("unigram unsmoothed: ",testword, res, "decimal:", pow2(res))

#Bigger test cases
test_case = "why dont we end there"
print("bigram completely seen sentence: ",test_case, wbSmooth(testlambs ,test_case, 2, debug=0))
#Trigrams test for fun
test_case = "why dont we end there"
print("trigram completely seen sentence: ",test_case, wbSmooth(testlambs ,test_case, 3, debug=0))
test_case = "let us start here"
print("trigram sentence: ",test_case, wbSmooth(testlambs ,test_case, 3, debug=0))

#print_dict(testlambs[0])
test_case = "let us start here"
print("bi sentence: ",test_case, wbSmooth(testlambs ,test_case, 2, debug=0))


unigram smoothed:  let 0.03125
uni_tokens, uni_types: 39 25
lambda.dot =  0.609375
unseen 0.015625
2 -grams: [(('<S_0>',), 'let'), (('let',), 'us'), (('us',), '</S_0>')]
current gram:  (('<S_0>',), 'let')
s:  [('<S_0>',), 'let'] ptemp:  0.125
mle [('<S_0>',), 'let'] 0.2
current gram:  (('let',), 'us')
s:  [('let',), 'us'] ptemp:  0.515625
mle [('let',), 'us'] 1.0
current gram:  (('us',), '</S_0>')
s:  [('us',), '</S_0>'] ptemp:  0.0078125
mle [('us',), '</S_0>'] 0
bigram smoothed:  let us 0.0005035400390625003
bigram unseen pair smoothed:  why why 1.9779911747685202e-05
bigram unseen pair smoothed:  us with 4.23855251736111e-06
bigram unseen tokens smoothed:  chicken chicken 1.6954210069444448e-06
unigram unsmoothed:  let 5.285402218862249 decimal: 0.025641025641025633
bigram completely seen sentence:  why dont we end there 0.004699843648941842
trigram completely seen sentence:  why dont we end there 0.01241116572933721
trigram sentence:  let us start here 0.0016515771403646793
bi sent

### Verifying small corpus by hand:
lambda.dot = tokens/(tokens+types) = 39/(39+25) = 39/64

Unigram "let" occurs once in the corpus of 39 tokens

therefore, P.wb(let) = (39/64) * (1/39) + (1/64) = **2/64**, 


consistent with the result above `unigram smoothed:  let 0.03125`

The Bigram of `<S_0>` let, is then equivalent to (5/9) * (1/5) + (4/9) * (2/64) = **0.125**
with lambda of `<S_0>` = 5/9
MLE(let|`<S_0>` ) = 1/5

consistent with the result above `s:  [('<S_0>',), 'let'] ptemp:  0.125`

### Analysis
The following test cases and outputs taken from above:

The results are as expected. As we move from a sentence with more bigrams in the corpus (e.g. for "let us", both 'S_0 let' AND 'let us' occur in the corpus) to one with 0-bigrams, but still with unigrams in the corpus ("us with"), to 0 sightings in the corpus at all ("chicken chicken"), the probability goes down.

    Input            Probability         Links
    let us           5.035 e-04          2/3 bigrams exist, 2/2 unigrams exist
    why why          1.977 e-05          1/3 bigrams exist, 2/2 unigrams exist
    us with          4.23 e-06           0/3 bigrams exist, 2/2 unigrams exist
    chicken chicken  1.69 e-06           0/3 bigrams exist, 0/2 unigrams exist

In [271]:
#LARGE CORPUS:

#wsj_corpus, wsj_norm_corpus
n = 3 #up to trigrams created
wsjlambs = []
make_lamb_dicts(wsjlambs, n, wsj_corpus, debug = 0)


In [274]:
#print_dict(wsjlambs[1])
test_case = "new"
print("unigram smoothed: ",test_case, wbSmooth(wsjlambs ,test_case, 1))
test_case = "new york"
print("bigram smoothed: ",test_case, wbSmooth(wsjlambs ,test_case, 2, debug=0))
test_case = "in new"
print("bigram smoothed: ",test_case, wbSmooth(wsjlambs ,test_case, 2, debug=0))



unigram smoothed:  new 0.002860885028626597
bigram smoothed:  new york 3.983696620112879e-05
bigram smoothed:  in new 5.314751430733009e-07


<font color="red">Self assessment:</font>

### Evaluation via Perplexity
Explore the effects of n-gram order using perplexity. Perform ten-fold cross-validation on the WSJ corpus. On each iteration, this will give you a different 90/10 training/test split; train a smoothed language model on the 9 training sections, and compute the average per-token perplexity of the tenth section. The slides from the language modeling lecture give the equation for perplexity computation (as does J&M chapter 4); you'll need to modify the equation a bit, since we're using log-probabilities. 

Now, try this for unigram, bigram, trigram, and 4-gram models. 

#### What to turn in
* your cross-validation function. You are not suppose to use any cross-validation function from any module. You should implement it by yourself.
* your perplexity function
* cross-validation result for unigram, bigram, trigram, and 4-gram models on wsj.pos.gz
* cross-validation result for unigram, bigram, trigram, and 4-gram models on wsj-normalized.pos.gz.
* Answer following 2 questions: 
    * How does perplexity change as the model order size increases?
    * How does perplexity change as the data changed?

In [301]:
def wbSmooth_fast(list_dicts, sentence, order, seen):
    """
    Faster WB smoothing function with dynamic programming
    Input:
        list_dicts: List of nested defaultdict(defaultdict(int)) structure containing n-gram counts
            from 1-gram to (order)-gram.
        sentence: Input sentence to calculate probability on
        order: degree of n-gram. Must be within provided (list_dict) range
    Output: LOG WB smoothed probability of input
    """
    uni_tokens = sum(list_dicts[0][()].values())
    uni_types = len(list_dicts[0][()].keys())
    lamb_dot = uni_tokens / (uni_tokens+uni_types)
    unseen = 1/(uni_tokens+uni_types)
    
    grams = list(ngrams([sentence.split()], order))

    probs = 0 #WB probabilities of component n-grams of the input

#Calculate smoothed prob for every n-gram in (sentence) and add to accumulator
    for gram in grams:
        prefix = gram[0]
        suffix = gram[1]

        ptemp =0.5

        for i in range(0,order):
            if gram in seen:
                ptemp = seen[gram]
                #print("seen n-gram", gram)
                break
            elif i == 0: #base case, unigrams
                if ((),suffix) in seen:
                    ptemp = seen[((),suffix)]
                    #print("seen unigram", suffix)
                else:
                    p = lamb_dot * (mle100(list_dicts[i], suffix)) + unseen
                    ptemp = p
                    seen[((),suffix)] = p
            else:
                gram_dict = list_dicts[i][prefix[order-i-1:]] #slice of prefix for current level of n-gram
                if (prefix[order-i-1:], suffix) in seen:
                    ptemp = seen[(prefix[order-i-1:], suffix)]
                    #print("seen partial-gram", prefix,suffix)
                else:
                    lamb = gram_dict.default_factory()
                    s = [prefix[order-i-1:], suffix]
                    mle = mle_solo(list_dicts[i], s)
                    ptemp = lamb*mle + (1-lamb)*ptemp
                    seen[(prefix[order-i-1:], suffix)] = ptemp
            
            #print("prefix", (gram[order-2-i],) ) #Unigram prefix
        probs += log2(ptemp)
    return probs

#Tests to see if match non-dynamic programming function
s = {}
test_case = "why dont we end there"
print("trigram completely seen sentence: ",test_case, pow2(wbSmooth_fast(testlambs ,test_case, 3,s)))
test_case = "let us start here"
print("trigram sentence: ",test_case, pow2(wbSmooth_fast(testlambs ,test_case, 3,s)) )
test_case = "let us start here"
print("bi sentence: ",test_case, pow2(wbSmooth_fast(testlambs ,test_case, 2,s)) )
test_case = "start start let let here"
print("bi sentence: ",test_case, pow2(wbSmooth_fast(testlambs ,test_case, 2,s)) )
#print(s)

#WSJ tests
s2 = {}
test_case = "new"
print("unigram smoothed: ",test_case, pow2(wbSmooth_fast(wsjlambs ,test_case, 1, s2)))
test_case = "new york"
print("bigram smoothed: ",test_case, pow2(wbSmooth_fast(wsjlambs ,test_case, 2, s2)))
test_case = "in new"
print("bigram smoothed: ",test_case, pow2(wbSmooth_fast(wsjlambs ,test_case, 2, s2)))

#Uni/Bigram/Trigram for same input
test_case = "in new york"
print("uni smoothed: ",test_case, pow2(wbSmooth_fast(wsjlambs ,test_case, 1, s2)))
test_case = "in new york"
print("bi smoothed: ",test_case, pow2(wbSmooth_fast(wsjlambs ,test_case, 2, s2)))
test_case = "in new york"
print("tri smoothed: ",test_case, pow2(wbSmooth_fast(wsjlambs ,test_case, 3, s2)))

test_case = "digital equipment corp. reported a 32 % decline in net income on a modest revenue gain in its fiscal first quarter  causing some analysts to predict weaker results ahead than they had expected"
ntemp = len(test_case.split()) + 2 #+ni, padding
pptemp = pow2( (-1/ntemp)*(wbSmooth_fast(wsjlambs ,test_case, 3, s2)) )
print("pptemp:", pptemp)

trigram completely seen sentence:  why dont we end there 0.01241116572933721
trigram sentence:  let us start here 0.0016515771403646793
bi sentence:  let us start here 0.003124382346868514
bi sentence:  start start let let here 1.1677462837578317e-09
unigram smoothed:  new 0.002860885028626597
bigram smoothed:  new york 3.983696620112879e-05
bigram smoothed:  in new 5.314751430733009e-07
uni smoothed:  in new york 5.45369236402621e-08
bi smoothed:  in new york 1.077026611164952e-05
tri smoothed:  in new york 4.374609091342155e-05
pptemp: 9.60149543497804


For the same input, "in new york", the probability increases going from 1-gram to 3-gram.

Thus, we should expect perplexity in the next section to decrease with respect to the order, N, as the
perplexity equation includes 1/P(WB)

In [ ]:
#from statistics import mean
#bigram padding +1 (ie. +(n-1) = +1)
#Using: wsj_corpus, wsj_norm_corpus
def PP(corpus, n):
    """ten-fold Cross validation from 1-gram to n-gram
    input:
        corpus: List of sentence strings
        n: order of n-gram to calculate"""
    #Split Data 9:1 training/testing chunks
    split = 10 #chunks of 10, 1/10th = test chunk
    perp = [[] for i in range(n)] #perplexity. [Size N list (size split list)]
    
    lc = len(corpus)
    for i in range(0,split): #test chunk different every loop
        train = corpus[0:i*lc//split] #starting chunk of training
        test = corpus[i*lc//split:(i+1)*lc//split] #test chunk in middle
        train += corpus[(i+1)*lc//split:lc] #add leftovers to training
        test_size = len(test)
        #Train models, 1-gram to n-gram in list of nested-defaultdicts
        train_lamb = []
        seen = {} #Seen n-grams probabilities
        make_lamb_dicts(train_lamb, n, train)
        
        #Test models, 1-gram to n-gram
        for ni in range(0,n):
            count = 0 #speed check
            pp_acc = 0 #sum of all per-token perplexity of each sentence, divide later by # sentences
            #Test each sentence
            for sentence in test:
                big_N = len(sentence.split()) + ni #+ni, padding
                #pp_acc += pow2( (-1/big_N)*(wbSmooth_fast(train_lamb ,sentence, ni+1, seen)) )
                pp_acc += (-1/big_N)*(wbSmooth_fast(train_lamb ,sentence, ni+1, seen)) #overflow? pow2 later
                #print(pp_acc, sentence)
                count += 1
                if count%100==0:
                    print(count)
            #pp_acc = (pp_acc / count)
            pp_acc = pow2(pp_acc / count)
            perp[ni] += [pp_acc] #store
            print(perp)
    return perp

perp1 = PP(wsj_corpus, 4)
perp2 = PP(wsj_norm_corpus, 4)
    

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558], [], [], []]
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558], [441.35754436281684], [], []]
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558], [441.35754436281684], [441.27562783514117], []]
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
20

2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558, 1650.5762357061963, 1534.5313768217015, 1564.7266215971847, 1581.4758587730742], [441.35754436281684, 461.68029077638363, 380.38064730342796, 393.4096670317867, 397.79106392650476], [441.27562783514117, 464.69316294604545, 362.75506759475195, 377.82773184280586, 377.716808532464], [585.5243809695246, 614.3222806450789, 474.1103591876566, 495.4393162129364]]
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558, 1650.5762357061963, 1534.5313768217015, 1564.7266215971847, 1581.4758587730742], [441.35754436281684, 461.68029077638363, 380.38064730342796, 393.4096670317867, 397.79106392650476], [441.27562783514117, 464.6931629460454

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558, 1650.5762357061963, 1534.5313768217015, 1564.7266215971847, 1581.4758587730742, 1562.3568473505134, 1577.1383920572532, 1561.6027529195692], [441.35754436281684, 461.68029077638363, 380.38064730342796, 393.4096670317867, 397.79106392650476, 378.3854354899437, 407.9718008996132, 393.4535041315459], [441.27562783514117, 464.69316294604545, 362.75506759475195, 377.82773184280586, 377.716808532464, 356.64721686803233, 393.2288058814871, 378.4270939968525], [585.5243809695246, 614.3222806450789, 474.1103591876566, 495.4393162129364, 492.64585922709136, 464.5134098177129, 517.5612220448373]]
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
290

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[1623.0199201636558, 1650.5762357061963, 1534.5313768217015, 1564.7266215971847, 1581.4758587730742, 1562.3568473505134, 1577.1383920572532, 1561.6027529195692, 1591.4282270191852, 1675.791494099099], [441.35754436281684, 461.68029077638363, 380.38064730342796, 393.4096670317867, 397.79106392650476, 378.3854354899437, 407.9718008996132, 393.4535041315459, 392.3027484037876, 416.6770600207211], [441.27562783514117, 464.69316294604545, 362.75506759475195, 377.82773184280586, 377.716808532464, 356.64721686803233, 393.2288058814871, 378.4270939968525, 374.9441078739071, 401.06358294509624], [585.5243809695246, 614.3222806450789, 474.1103591876566, 495.4393162129364, 492.64585922709136, 464.5134098177129, 517.5612220448373, 498.9879115321822, 

100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[652.0887869743311, 632.5453401659892, 571.5914141545143, 584.3991471977047, 600.5520430157416], [210.67377561147742, 207.68307643100707, 174.80321192934804, 179.5726119954854, 183.545610612244], [198.60331710261283, 196.24440162764063, 158.6940645018769, 162.79671312078858], [248.82179995749175, 245.02569538996684, 195.80409845590654, 200.74056502879915]]
100
200
300
400
500
600
700
800
900
1000
1100
1200
1300
1400
1500
1600
1700
1800
1900
2000
2100
2200
2300
2400
2500
2600
2700
2800
2900
3000
3100
3200
3300
3400
3500
3600
3700
3800
3900
4000
4100
4200
4300
4400
4500
4600
4700
4800
4900
5000
5100
5200
[[652.0887869743311, 632.5453401659892, 571.5914141545143, 584.3991471977047, 600.5520430157416], [210.67377561147742, 207.68307643100707,

In [269]:
print("Cross validation: Perplexity per token")
print("perp wsj, 1-gram to 4-gram", [sum(elem)/len(elem) for elem in perp1])
print("perp norm", [sum(elem)/len(elem) for elem in perp2])

Cross validation: Perplexity per token
perp wsj, 1-gram to 4-gram [4.3436506259667995, 4.5840537141539786, 6.09661813239501, 8.856099690034466]
perp norm [4.602251067371387, 4.974512352715598, 6.372987318580018, 8.780928008682977]


In [292]:
#Issue[1]: Perplexities too low
#Bug in p4 above... used len (characters) instead of len(words)
print((wsj_corpus[0]))
print(len(wsj_corpus[0]))
print(len(wsj_corpus[0].split()))

#New Issue[2]: Perplexity too high

digital equipment corp. reported a 32 % decline in net income on a modest revenue gain in its fiscal first quarter  causing some analysts to predict weaker results ahead than they had expected
192
33


## Results 0.1 (too low)
Cross validation: Perplexity per token
perp wsj, 1-gram to 4-gram [4.3436506259667995, 4.5840537141539786, 6.09661813239501, 8.856099690034466]
perp norm [4.602251067371387, 4.974512352715598, 6.372987318580018, 8.780928008682977]

ISSUE: Using character length of sentence instead of word length
FIX: `big_N = len(sentence) + ni`->`big_N = len(sentence.split()) + ni`

## Results 0.2 (too high)
[[19337.134726670945], [36032.115546286026], [31975.143858626707], [31454.97475573105]]
[[3749.846972451751], [5978.871676394404], [5449.661015094365], [5877.729005236297]] <- lowercased?
Issue: Algorithm? Log?

Overflow?
`pp_acc += pow2( (-1/big_N)*(wbSmooth_fast(train_lamb ,sentence, ni+1, seen)) )`
->                `pp_acc += (-1/big_N)*(wbSmooth_fast(train_lamb ,sentence, ni+1, seen))`

MOVED pow2() till later after averaging WB_all() to prevent potential overflow?

## Results Final
WSJ: [[1623.0199201636558, 1650.5762357061963, 1534.5313768217015, 1564.7266215971847, 1581.4758587730742, 1562.3568473505134, 1577.1383920572532, 1561.6027529195692], [441.35754436281684, 461.68029077638363, 380.38064730342796, 393.4096670317867, 397.79106392650476, 378.3854354899437, 407.9718008996132, 393.4535041315459], [441.27562783514117, 464.69316294604545, 362.75506759475195, 377.82773184280586, 377.716808532464, 356.64721686803233, 393.2288058814871, 378.4270939968525], [585.5243809695246, 614.3222806450789, 474.1103591876566, 495.4393162129364, 492.64585922709136, 464.5134098177129, 517.5612220448373, 498.9879115321822]]
WSJ: 1580, 400, 390, 515
WSJ_Norm: 600, 190, 170, 220

The general trend for perplexity decreases as order increases, which makes sense given our previous calculations showing that probability increases with order. Since perplexity contains a 1/P term, as order increases, probability increases, and perplexity decreases.

The WSJ_norm perplexity is lower compared to the regular version. This can be attributed to the aggregation of rare terms such as proper nouns, resulting in a greater match, as a test case containing "in City, State" would be converted to "in <Proper Noun> <Proper Noun>". As calculated above in part 2, the wsj bigram models have 43754 keys for the regular-model, and 29006 keys for the norm-model -- a significant reduction.
    
The final unexpected pattern, in that the 4th-order perplexity actually was greater than the 2nd, 3rd perplexities for both models, could be due to insufficient data as Meysam mentioned in class. A sufficiently large corpus would contain the vast majority of n-grams in the test-set, but it's possible our training set did not contain enough 4-grams in the test set to continue the downward trend in perplexity.


<font color="red">Self assessment:</font>